caddyhttp/proxy/reverseproxy.go

```diff
@@ -119,12 +130,38 @@ func NewSingleHostReverseProxy(target *url.URL, without string, keepalive int) *
 			}
 		}
 
-		hadTrailingSlash := strings.HasSuffix(req.URL.Path, "/")
-		req.URL.Path = path.Join(target.Path, req.URL.Path)
-		// path.Join will strip off the last /, so put it back if it was there.
-		if hadTrailingSlash && !strings.HasSuffix(req.URL.Path, "/") {
-			req.URL.Path = req.URL.Path + "/"
+		var a, b string
+		if req.URL.Opaque != "" || target.Opaque != "" {
+			if target.Opaque != "" {
+				a = target.Opaque
+			} else {
+				a = target.Path
+			}
+
+			if req.URL.Opaque != "" {
+				b = req.URL.Opaque
+			} else {
+				b = req.URL.Path
+			}
+
+			req.URL.Opaque = singleJoiningSlash(a, b)
+		}
+		if req.URL.RawPath != "" || target.RawPath != "" {
+			if target.RawPath != "" {
+				a = target.RawPath
+			} else {
+				a = target.Path
+			}
+
+			if req.URL.RawPath != "" {
+				b = req.URL.RawPath
+			} else {
+				b = req.URL.Path
+			}
+
+			req.URL.RawPath = singleJoiningSlash(a, b)
```

Comment:
I think I'm OK with this logic, but could you comment the relevant parts, like why the `if` statements are the way they are, or what the point of these `if` statements is? Maybe this chunk belongs in a separate function.